tracer/src/Datadog.Trace/Agent/MessagePack/SpanMessagePackFormatter.cs

```diff
@@ -189,9 +197,57 @@ private int Serialize(ref byte[] bytes, int offset, in SpanModel spanModel)
             offset += WriteTags(ref bytes, offset, in spanModel, tagProcessors);
             offset += WriteMetrics(ref bytes, offset, in spanModel, tagProcessors);
 
+            if (metaStructValues?.Count > 0)
+            {
+                WriteMetaStruct(ref bytes, ref offset, metaStructValues);
+            }
+
             return offset - originalOffset;
         }
 
+        private void WriteMetaStruct(ref byte[] bytes, ref int offset, IReadOnlyDictionary<string, object> metaStruct)
+        {
+            var data = new Dictionary<string, byte[]>();
+
+            // Since ConcurrentDictionary does not guarantee the order of the keys, we need to order them
+            // to make serialization deterministic.
+            // We need to know the size of the byte array before actually writing it. Depending on the size,
+            // the serializer will use a different amount of bytes in the header to encode the length of the byte array.
+            // That means that we need to know the size of the array, then write the header and then, the array itself.
+
+            foreach (var keyValuePair in metaStruct.OrderBy(keyValuePair => keyValuePair.Key, StringComparer.OrdinalIgnoreCase))
+            {
+                // 256 is the size that the serializer would reserve initially for empty arrays, so we create
+                // the buffer with that size to avoid this first resize. If a bigger size is required later, the serializer
+                // will resize it.
+                var buffer = new byte[256];
+
+                // PrimitiveObjectFormatter relies on MessagePackBinary
+                var bytesCopied = PrimitiveObjectFormatter.Instance.Serialize(ref buffer, 0, keyValuePair.Value, null);
+
+                if (bytesCopied != buffer.Length)
+                {
+                    Array.Resize(ref buffer, bytesCopied);
```

Comment:
Doesn't `PrimitiveObjectFormatter.Instance.Serialize()` already resize this buffer if needed?

Reply:
Yes, we don't want to send more bytes than the ones used. bytesCopied should not be bigger than buffer.length. if it is, then we have copied bytes out of the bounds of buffer.

Reply:
Thanks for clarifying.